src/models/ModelBuilder.cc

```diff
@@ -3,23 +3,136 @@
+        builtDefinitions[symbolSigVal.first] = std::move(templateFun);
+    }
+    return std::unique_ptr<Model>(new Model(logic, assignment, builtDefinitions));
+}
+
+void ModelBuilder::addToTheoryFunction(SymRef sr, vec<PTRef> vals, PTRef val)
+{
+    if (logic.getSortRef(sr) != logic.getSortRef(val)) {
+        throw OsmtApiException("Incompatible sort for symbol valuation");
+    }
+    Symbol & sym = logic.getSym(sr);
+    if (sym.nargs() != vals.size_()) {
+        throw OsmtApiException("Incorrect valuation for symbol: argument and valuation size do not match");
+    }
+    for (int i = 0; i < vals.size(); i++) {
+        if (sym[i] != logic.getSortRef(vals[i])) {
+            throw OsmtApiException("Incorrect valuation for symbol: sort mismatch");
+        }
+    }
+
+    if (not hasTheoryFunction(sr)) {
+        vec<PTRef> formalArgs; formalArgs.capacity(vals.size());
+        std::string symName(logic.getSymName(sr));
+        // Ensure that no formal arg name collides with the function name
+        std::string formalArgPrefix(Model::getFormalArgBaseNameForSymbol(logic, sr, formalArgDefaultPrefix));
+
+        for (PTRef v : vals) {
+            std::stringstream ss;
+            ss << formalArgPrefix << uniqueNum++;
+            formalArgs.push(logic.mkVar(logic.getSortRef(v), ss.str().c_str()));
+        }
+        FunctionSignature templateSig(logic.protectName(logic.getSymName(sr)), std::move(formalArgs), logic.getSortRef(sr));
+        definitions.insert({sr,opensmt::pair<FunctionSignature,ValuationNode*>{std::move(templateSig), nullptr}});
+    }
+    auto & signatureAndValuation = definitions.at(sr);
+    FunctionSignature & templateSignature = signatureAndValuation.first;
+    const vec<PTRef> & formalArgs = templateSignature.getArgs();
+    vec<opensmt::pair<PTRef,PTRef>> valuation; valuation.capacity(vals.size());
+    for (int i = 0; i < vals.size(); i++) {
+        valuation.push({formalArgs[i], vals[i]});
+    }
+    assert(logic.getSortRef(val) == logic.getSortRef(sr));
+    signatureAndValuation.second = addToValuationTree(valuation, val, signatureAndValuation.second);
+}
+
+ModelBuilder::ValuationNode * ModelBuilder::addToValuationTree(const vec<opensmt::pair<PTRef,PTRef>> & valuation, PTRef value, ValuationNode * root) {
+    assert(valuation.size() >= 1);
+
+    if (root == nullptr) {
+        root = valuationNodeFactory.getValuationNode(valuation[0].first, PTRef_Undef);
+    }
+
+    ValuationNode * current = root;
+
+    for (int i = 0; i < valuation.size(); i++) {
+        auto & varVal = valuation[i];
+        PTRef val = varVal.second; // The value of my parent's var
+
+        assert(current != nullptr);
+
+        bool foundChild = false;
+        for (auto child: *current) {
+            if (child->getValue() == val) {
+                current = child;
+                foundChild = true;
+                break;
+            }
+        }
+        if (foundChild) {
```

Comment:
This control-flow can be simplified, no?

Reply:
OK.